modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java

```diff
@@ -113,30 +116,39 @@
             builder.indexAnalyzer(parserContext.getIndexAnalyzers().getDefaultIndexAnalyzer());
             builder.searchAnalyzer(parserContext.getIndexAnalyzers().getDefaultSearchAnalyzer());
             builder.searchQuoteAnalyzer(parserContext.getIndexAnalyzers().getDefaultSearchQuoteAnalyzer());
-            parseTextField(builder, name, node, parserContext);
             for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
                 final Map.Entry<String, Object> entry = iterator.next();
                 final String fieldName = entry.getKey();
                 final Object fieldNode = entry.getValue();
-
+                checkNull(fieldName, fieldNode);
                 if (fieldName.equals("max_shingle_size")) {
                     builder.maxShingleSize(nodeIntegerValue(fieldNode));
                     iterator.remove();
+                } else if (fieldName.equals("similarity")) {
+                    SimilarityProvider similarityProvider = TypeParsers.resolveSimilarity(parserContext, fieldName, fieldNode.toString());
+                    builder.similarity(similarityProvider);
+                    iterator.remove();
                 }
                 // TODO should we allow to configure the prefix field
             }
+            parseTextField(builder, name, node, parserContext);
```

Comment:
The shared `parseTextField()` method does a deprecation check for `similarity` now, so we need to move it to after we've processed the `similarity` field.  To ensure that mapper-specific fields still get checked for `null`, a new `TypeParsers.checkNull()` method is called for each entry.